sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala

```diff
@@ -1151,6 +1153,21 @@ class QueryExecutionErrorsSuite
       )
     )
   }
+
+  test("SPARK-43259: Uses unsupported custom encoder") {
+    class CustomEncoder extends Encoder[Int] {
+      override def schema: StructType = StructType(Array.empty[StructField])
+      override def clsTag: ClassTag[Int] = ClassTag.Int
+    }
+    val e = intercept[SparkRuntimeException] {
+      encoderFor(new CustomEncoder)
```

Comment:
Can you try to reproduce the issue using public Spark SQL API (sql or Java/Scala/Python api)? If it is not possible, we should convert the error to an internal one.

Reply:
I explored the problem and have changed the PR description accordingly, could you take a look at it again, if it makes sense to you as well? @MaxGekk

Reply:
I think this is an user facing error, see how to reproduce it using public API:
```scala
    implicit val kryoEncoder = new Encoder[KryoData] {
      override def schema: StructType = StructType(Array.empty[StructField])
      override def clsTag: ClassTag[KryoData] = ???
    }
    val ds = Seq(KryoData(1), KryoData(2)).toDS()
    ds.groupByKey(p => p).count().collect()
```
it fails with:
```java
Only expression encoders are supported for now.
org.apache.spark.SparkRuntimeException: Only expression encoders are supported for now.
	at org.apache.spark.sql.errors.QueryExecutionErrors$.unsupportedEncoderError(QueryExecutionErrors.scala:463)
	at org.apache.spark.sql.catalyst.encoders.package$.encoderFor(package.scala:34)
```

Reply:
I have introduced this change. Just out of curiosity, should we change the error format to pass the type of encoder that was used and then include it in the message?